src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.SyntaxAndDeclarationManager.cs

```diff
@@ -0,0 +1,587 @@
+
+                var treesBuilder = ArrayBuilder<SyntaxTree>.GetInstance();
+                var ordinalMapBuilder = PooledDictionary<SyntaxTree, int>.GetInstance();
+                var declMapBuilder = this.RootNamespaces.ToBuilder();
+                var declTable = this.DeclarationTable;
+                foreach (var tree in syntaxTrees)
+                {
+                    if (removeSet.Contains(tree))
+                    {
+                        loadDirectiveMap = loadDirectiveMap.Remove(tree);
+                        loadedSyntaxTreeMap = loadedSyntaxTreeMap.Remove(tree.FilePath);
+                        RemoveSyntaxTreeFromDeclarationMapAndTable(tree, declMapBuilder, ref declTable);
+                        if (tree.HasReferenceDirectives())
+                        {
+                            reuseReferenceManager = false;
+                        }
+                    }
+                    else if (!loadedSyntaxTreeMap.ContainsKey(tree.FilePath))
+                    {
+                        UpdateSyntaxTreesAndOrdinalMapOnly(
+                            treesBuilder,
+                            tree,
+                            ordinalMapBuilder,
+                            loadDirectiveMap,
+                            loadedSyntaxTreeMap);
+                    }
+                }
+
+                return Update(
+                    treesBuilder.ToImmutableAndFree(),
+                    ordinalMapBuilder.ToImmutableDictionaryAndFree(),
+                    loadDirectiveMap,
+                    loadedSyntaxTreeMap,
+                    declMapBuilder.ToImmutableDictionary(),
+                    declTable);
+            }
+
+            /// <summary>
+            /// Collects all the trees #load'ed by <paramref name="oldTree"/> (as well as
+            /// <paramref name="oldTree"/> itself) and populates <paramref name="removeSet"/>
+            /// with all the trees that are safe to remove (not #load'ed by any other tree).
+            /// </summary>
+            private static void GetRemoveSet(
+                SyntaxTree oldTree,
+                bool includeLoadedTrees,
+                ImmutableArray<SyntaxTree> syntaxTrees,
+                ImmutableDictionary<SyntaxTree, int> syntaxTreeOrdinalMap,
+                ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+                ImmutableDictionary<string, SyntaxTree> loadedSyntaxTreeMap,
+                HashSet<SyntaxTree> removeSet,
+                out int totalReferencedTreeCount,
+                out ImmutableArray<LoadDirective> oldLoadDirectives)
+            {
+                // Check to make sure this is not a #load'ed tree.
+                SyntaxTree loadedTree;
+                if (loadedSyntaxTreeMap.TryGetValue(oldTree.FilePath, out loadedTree))
+                {
+                    throw new ArgumentException(string.Format(CSharpResources.SyntaxTreeFromLoadNoRemoveReplace, loadedTree), "trees");
```

Comment:
RE: 'oldTree' vs 'loadedTree'...they *should* be the same, but I should add an explicit check for that...

RE: ```nameof(parameter)```, this is actually a parameter of the caller...maybe this check should be moved...